finality-aleph/src/sync/handler.rs

```diff
@@ -265,22 +227,45 @@ impl<I: PeerId, J: Justification, CS: ChainStatus<J>, V: Verifier<J>, F: Finaliz
             .verifier
             .verify(state.top_justification())
             .map_err(Verifier)?;
+        let remote_top_number = remote_top.header().id().number();
         let local_top = self.chain_status.top_finalized().map_err(ChainStatus)?;
-        match local_top
-            .header()
-            .id()
-            .number()
-            .checked_sub(remote_top.header().id().number())
-        {
-            // If we are just one behind then normal broadcasts should remedy the situation.
-            Some(0..=1) => Ok(SyncActions::noop()),
-            Some(2..=MAX_SMALL_JUSTIFICATION_BATCH) => Ok(SyncActions::response(
-                self.small_justification_batch_from(remote_top.header().id())?,
+        let local_top_number = local_top.header().id().number();
+        let remote_session = session_id_from_block_num(remote_top_number, self.period);
+        let local_session = session_id_from_block_num(local_top_number, self.period);
+        if remote_top_number >= local_top_number {
+            return self.handle_verified_justification(remote_top, peer);
+        }
+        match local_session.0.checked_sub(remote_session.0) {
+            Some(0) => Ok(SyncActions::state_broadcast_response(
+                local_top.into_unverified(),
+                None,
+            )),
+            Some(1) => Ok(SyncActions::state_broadcast_response(
+                self.chain_status
```

Comment:
This big block of code is called three times here. Maybe it would be nice to have a function like 
```
fn last_finalized(&self, session) -> Ok<J::Unverified, Error<J, CS, V, F>> {...}
```

Reply:
well, **technically** this is a single line of code...

Reply:
(fixed)